system-addon/lib/MessageCenterRouter.jsm

```diff
@@ -0,0 +1,204 @@
+const {actionTypes: at} = ChromeUtils.import("resource://activity-stream/common/Actions.jsm", {});
+ChromeUtils.import("resource://gre/modules/Console.jsm");
+ChromeUtils.import("resource://gre/modules/Timer.jsm");
+
+const INCOMING_MESSAGE_NAME = "MessageCenter:child-to-parent";
+const OUTGOING_MESSAGE_NAME = "MessageCenter:parent-to-child";
+
+const FAKE_MESSAGES = [
+  {
+    id: "ONBOARDING_1",
+    template: "simple_snippet",
+    content: {
+      title: "Find it faster",
+      body: "Access all of your favorite search engines with a click. Search the whole Web or just one website from the search box.",
+      button: {
+        label: "Learn More",
+        action: "OPEN_LINK",
+        params: {url: "https://mozilla.org"}
+      }
+    }
+  },
+  {
+    id: "ONBOARDING_2",
+    template: "simple_snippet",
+    content: {
+      title: "Make Firefox your go-to-browser",
+      body: "It doesn't take much to get the most from Firefox. Just set Firefox as your default browser and put control, customization, and protection on autopilot."
+    }
+  },
+  {
+    id: "ONBOARDING_3",
+    template: "simple_snippet",
+    content: {
+      title: "Did you know?",
+      body: "All human beings are born free and equal in dignity and rights. They are endowed with reason and conscience and should act towards one another in a spirit of brotherhood."
+    }
+  }
+];
+
+/**
+ * getRandomItemFromArray
+ *
+ * @param {Array} arr An array of items
+ * @returns one of the items in the array
+ */
+function getRandomItemFromArray(arr) {
+  const index = Math.floor(Math.random() * arr.length);
+  return arr[index];
+}
+
+/**
+ * @class _MessageCenterRouter - Keeps track of all messages, UI surfaces, and
+ * handles blocking, rotation, etc. Inspecting MessageCenter.state will
+ * tell you what the current displayed message is in all UI surfaces.
+ *
+ * Note: This is written as a constructor rather than just a plain object
+ * so that it can be more easily unit tested.
+ */
+class _MessageCenterRouter {
+  constructor(initialState = {}) {
+    this.initialized = false;
+    this.messageChannel = null;
+    this._state = Object.assign({
+      currentId: null,
+      blockList: {},
```

Comment:
Right now all the consumers of indexedDB are feeds. The object stores are created in `ActivityStream.jsm` before the feeds are initialized and then each feed creates it's own `_storage` that connects to the required store.
If we were to connect to indexedDB in the Router here (to fetch the blockList) it might happen before AS.jsm (maybe?).
I think creating the router should depend on the feed so that it's easier to pass in indexedDB data.

Reply:
why in `init`? initial state doesn't cause any side effects

Reply:
So that we can initialize the state with the blockList values stored.